File: src/ladym/langgraph/tools.py
"""Path A — LangChain/LangGraph tools wrapping ladyM long-term memory.

Use :func:`create_ladym_tools` when the LLM should decide when to recall /
remember (ReAct-style). The workspace is bound at factory time; for per-request
multi-user isolation use :mod:`ladym.langgraph.nodes` instead.
"""
from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

from ..config import Config
from ..engine import Engine
from ._runtime import resolve_engine

if TYPE_CHECKING:
    from langchain_core.tools import BaseTool


def create_ladym_tools(
    engine: Engine | Config | str | Path | None = None,
    *,
    workspace: str | None = None,
    default_top_k: int = 8,
) -> list[BaseTool]:
    """Build LangChain tools backed by ladyM memory.

    Returns ``[recall_memory, remember_fact, search_code]``. Each tool closes
    over one Engine + workspace resolved at factory time. ``workspace`` is
    honored only when ``engine`` is NOT an already-constructed ``Engine``
    (an Engine's own ``config.workspace`` wins).
    """
    from langchain_core.tools import tool  # lazy: not a hard dep of ladym core

    eng = resolve_engine(engine, workspace=workspace)
    ws = eng.config.workspace

    @tool
    def recall_memory(query: str, top_k: int = default_top_k) -> str:
        """Retrieve relevant long-term memories (facts, decisions, code, playbooks).

        Call BEFORE answering when prior context, past decisions, or how-to
        knowledge may be relevant. One line per hit, or "(no hits)".
        """
        resp = eng.recall(query, workspace=ws, top_k=top_k)
        if not resp.results:
            return "(no hits)"
        return "\n".join(
            f"[{r.memory.layer}|{r.memory.type}|{r.score:.2f}] "
            f"{r.memory.summary or r.memory.content}"
            for r in resp.results
        )

    @tool
    def remember_fact(content: str, tags: list[str] | None = None) -> str:
        """Persist a durable fact worth recalling later (user preference, key
        decision, verified answer). Do NOT store ephemeral/transactional state.

        Subject to ladyM's attention gate: low-value or duplicate content may be
        dropped or rewritten automatically.
        """
        mem = eng.remember(content, tags=tags or [], source="langgraph-tool")
        gate = mem.metadata.get("gated", "pass")
        return f"stored id={mem.id} gate={gate}"

    @tool
    def search_code(query: str, top_k: int = default_top_k) -> str:
        """Search indexed source symbols/files. Use for 'how does X work in the
        codebase' against previously ``index_code``-indexed source.
        """
        resp = eng.search_code(query, workspace=ws, top_k=top_k)
        if not resp.results:
            return "(no hits)"
        return "\n".join(
            f"{r.memory.summary} :: {r.memory.metadata.get('qualified_name', '')}"
            for r in resp.results
        )

    return [recall_memory, remember_fact, search_code]

File: langgraph/tools.go
// Package langgraph integrates ladyM as a memory layer for langchain-golang
// LangGraph hosts — the Go port of the Python langgraph module (main branch).
//
// Two paths, mirroring the Python module:
//
//   - Path A (Tools): CreateTools returns LangChain tools wrapping ladyM
//     long-term memory; the LLM decides when to recall / remember
//     (ReAct-style, e.g. via agents.CreateAgent).
//   - Path B (Nodes): CreateRecallNode / CreateRetainNode return graph
//     NodeFuncs for automatic per-turn memory injection, wired into the
//     host's own StateGraph.
package langgraph

import (
	"context"
	"fmt"
	"strings"

	"github.com/ProjAnvil/LadyM/engine"
	"github.com/ProjAnvil/LadyM/schema"
	lctools "github.com/projanvil/langchain-golang/core/tools"
)

// CreateTools builds LangChain tools backed by ladyM memory:
// [recall_memory, remember_fact, search_code]. Each tool closes over the
// given Engine; workspace overrides the engine's default workspace (empty
// keeps the engine default). Mirrors Python's create_ladym_tools.
func CreateTools(eng *engine.Engine, workspace string, defaultTopK int) ([]lctools.Tool, error) {
	if defaultTopK <= 0 {
		defaultTopK = 8
	}
	ws := workspace
	if ws == "" {
		ws = eng.Config.Workspace
	}

	type recallArgs struct {
		Query string `json:"query"`
		TopK  int    `json:"top_k,omitempty"`
	}
	type rememberArgs struct {
		Content string   `json:"content"`
		Tags    []string `json:"tags,omitempty"`
	}

	recallMemory, err := lctools.FromFunc("recall_memory",
		"Retrieve relevant long-term memories (facts, decisions, code, playbooks).\n\n"+
			"Call BEFORE answering when prior context, past decisions, or how-to "+
			"knowledge may be relevant. One line per hit, or \"(no hits)\".",
		func(ctx context.Context, in recallArgs) (lctools.Result, error) {
			topK := in.TopK
			if topK <= 0 {
				topK = defaultTopK
			}
			resp, err := eng.Recall(in.Query, ws, topK, nil, nil, 0)
			if err != nil {
				return lctools.Result{}, err
			}
			if len(resp.Results) == 0 {
				return lctools.Result{Content: "(no hits)"}, nil
			}
			lines := make([]string, 0, len(resp.Results))
			for _, r := range resp.Results {
				summary := r.Memory.Summary
				if summary == "" {
					summary = r.Memory.Content
				}
				lines = append(lines, fmt.Sprintf("[%s|%s|%.2f] %s", r.Memory.Layer, r.Memory.Type, r.Score, summary))
			}
			return lctools.Result{Content: strings.Join(lines, "\n")}, nil
		})
	if err != nil {
		return nil, err
	}

	rememberFact, err := lctools.FromFunc("remember_fact",
		"Persist a durable fact worth recalling later (user preference, key "+
			"decision, verified answer). Do NOT store ephemeral/transactional state.\n\n"+
			"Subject to ladyM's attention gate: low-value or duplicate content may "+
			"be dropped or rewritten automatically.",
		func(ctx context.Context, in rememberArgs) (lctools.Result, error) {
			m, err := eng.Remember(in.Content, schema.LayerSemantic, schema.TypeFact, in.Tags, nil, "langgraph-tool", "")
			if err != nil {
				return lctools.Result{}, err
			}
			gate, _ := m.Metadata["gated"].(string)
			if gate == "" {
				gate = "pass"
			}
			return lctools.Result{Content: fmt.Sprintf("stored id=%s gate=%s", m.ID, gate)}, nil
		})
	if err != nil {
		return nil, err
	}

	searchCode, err := lctools.FromFunc("search_code",
		"Search indexed source symbols/files. Use for 'how does X work in the "+
			"codebase' against previously index_code-indexed source.",
		func(ctx context.Context, in recallArgs) (lctools.Result, error) {
			topK := in.TopK
			if topK <= 0 {
				topK = defaultTopK
			}
			resp, err := eng.SearchCode(in.Query, topK, ws)
			if err != nil {
				return lctools.Result{}, err
			}
			if len(resp.Results) == 0 {
				return lctools.Result{Content: "(no hits)"}, nil
			}
			lines := make([]string, 0, len(resp.Results))
			for _, r := range resp.Results {
				qn, _ := r.Memory.Metadata["qualified_name"].(string)
				lines = append(lines, fmt.Sprintf("%s :: %s", r.Memory.Summary, qn))
			}
			return lctools.Result{Content: strings.Join(lines, "\n")}, nil
		})
	if err != nil {
		return nil, err
	}

	return []lctools.Tool{recallMemory, rememberFact, searchCode}, nil
}
